aea/agent_loop.py

```diff
@@ -0,0 +1,402 @@
+
+    NEW_BEHAVIOURS_PROCESS_SLEEP = 1  # check new behaviours registered every second.
+
+    def __init__(self, agent: "AEA", loop: AbstractEventLoop = None):
+        """
+        Init agent loop.
+
+        :param agent: AEA instance
+        :param loop: asyncio loop to use. optional
+        """
+        super().__init__(agent)
+        self._agent: "AEA" = self._agent
+
+        try:
+            self._loop = loop or asyncio.get_event_loop()
+            assert not self._loop.is_closed()
+            assert not self._loop.is_running()
+        except (RuntimeError, AssertionError):
+            self._loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(self._loop)
+
+        self._behaviours_registry: Dict[TickerBehaviour, PeriodicCaller] = {}
+        self._state: AsyncState = AsyncState()
+        self._exceptions: List[Exception] = []
+
+    def start(self):
+        """Start agent loop."""
+        self._state.set("started")
+        self._loop.run_until_complete(self._run())
+
+    def stop(self):
+        """Stop agent loop."""
+        self._state.set("stopping")
+
+    def _behaviour_exception_callback(self, fn: Callable, exc: Exception) -> None:
+        """
+        Call on behaviour's act exception.
+
+        :param fn: behaviour's act
+        :param: exc: Exception  raised
+
+        :return: None
+        """
+        logger.exception(f"Exception: `{exc}` occured during `{fn}` processing")
+        self._exceptions.append(exc)
+        self._state.set("error")
+
+    def _register_behaviour(self, behaviour: TickerBehaviour) -> None:
+        """
+        Register behaviour to run periodically.
+
+        :param behaviour: TickerBehaviour object
+
+        :return: None
+        """
+        if behaviour in self._behaviours_registry:
+            # already registered
+            return
+
+        periodic_caller = PeriodicCaller(
+            behaviour.act,
+            behaviour._tick_interval,
+            behaviour._start_at,
+            self._behaviour_exception_callback,
+            self._loop,
+        )
+        self._behaviours_registry[behaviour] = periodic_caller
+        periodic_caller.start()
+
+    def _register_all_behaviours(self) -> None:
+        """Register all AEA behaviours to run periodically."""
+        for behaviour in self._agent._get_active_behaviours():
+            if not isinstance(behaviour, TickerBehaviour):
+                logger.error(f"{self.__class__} supports only TieckerBehaviour")
+                continue
+            self._register_behaviour(behaviour)
+
+    def _unregister_behaviour(self, behaviour: TickerBehaviour) -> None:
+        """
+        Unregister periodic execution of the behaviour.
+
+        :param behaviour: TockerBehaviour to schedule periodic execution.
+        :return: None
+        """
+        periodic_caller = self._behaviours_registry.pop(behaviour, None)
+        if not periodic_caller:
```

Comment:
prefer explicit `if periodic_caller is None:`

Reply:
+